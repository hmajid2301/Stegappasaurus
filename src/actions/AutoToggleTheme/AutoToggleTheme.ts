import AsyncStorage from "@react-native-community/async-storage";
import * as Location from "expo-location";
import * as Permissions from "expo-permissions";
import { getSunrise, getSunset } from "sunrise-sunset-js";

import Snackbar from "~/actions/Snackbar";

export default class AutoToggleTheme {
  private static oneDay = 24 * 60 * 60 * 1000;

  public async shouldToggleDarkTheme() {
    const currentTime = new Date(Date.now());
    const { sunrise, sunset } = await this.getSunriseAndSunsetTime(currentTime);
    let toggleTheme = true;

    if (sunrise !== null && sunset !== null) {
      if (currentTime > sunrise && currentTime < sunset) {
        toggleTheme = false;
      }
    }

    return toggleTheme;
  }

  private async getSunriseAndSunsetTime(currentTime: Date) {
    const { latitude, longitude } = await this.getLatitudeLongitude();
    let sunset = null;
    let sunrise = null;

    if (!isNaN(latitude)) {
      sunset = getSunset(latitude, longitude, currentTime);
      sunrise = getSunrise(latitude, longitude, currentTime);

      if (sunrise > sunset) {
        sunset = new Date(sunset.getTime() + AutoToggleTheme.oneDay);
      }
    }
    return { sunset, sunrise };
  }

  private async getLatitudeLongitude() {
    const currentDate = new Date(Date.now());
    const lastQueried = await AsyncStorage.getItem("@LastQueriedLocation");
    let latitude: number;
    let longitude: number;
    let lastQueriedDate: Date;

    if (lastQueried) {
      lastQueriedDate = new Date(lastQueried);
    } else {
      lastQueriedDate = new Date(
        currentDate.getTime() - AutoToggleTheme.oneDay
      );
    }

    if (
      currentDate.getTime() - lastQueriedDate.getTime() >=
      AutoToggleTheme.oneDay
    ) {
      ({ latitude, longitude } = await this.getNewLatitudeLongitude(
        currentDate
      ));
    } else {
      latitude = Number((await AsyncStorage.getItem("@Latitude")) || "51.5");
      longitude = Number((await AsyncStorage.getItem("@Longitude")) || "-0.12");
    }

    return { latitude, longitude };
  }

  private async getNewLatitudeLongitude(currentDate: Date) {
    let latitude;
    let longitude;
    const { status } = await Permissions.askAsync(Permissions.LOCATION);
    if (status === "granted") {
      const locationOn = await Location.hasServicesEnabledAsync();

      if (locationOn) {
        const location = await Location.getCurrentPositionAsync({});
        latitude = location.coords.latitude;
        longitude = location.coords.longitude;

        await Promise.all([
          AsyncStorage.setItem("@Latitude", JSON.stringify(latitude)),
          AsyncStorage.setItem("@Longitude", JSON.stringify(longitude)),
          AsyncStorage.setItem(
            "@LastQueriedLocation",
            JSON.stringify(currentDate)
          )
        ]);
      }
    }

    if (latitude === undefined || longitude === undefined) {
      throw Error("No location found");
    }

    return { latitude, longitude };
  }
}
