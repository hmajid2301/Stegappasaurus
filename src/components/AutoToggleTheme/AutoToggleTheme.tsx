import * as Location from "expo-location";
import * as Permissions from "expo-permissions";
import moment from "moment";
import { AsyncStorage } from "react-native";
import { getSunrise, getSunset } from "sunrise-sunset-js";

import Snackbar from "~/components/Snackbar";

export default class AutoToggleTheme {
  public shouldToggleTheme = async () => {
    const { sunrise, sunset } = await this.getSunriseAndSunsetTime();
    let toggleTheme = false;

    if (sunrise !== null && sunset !== null) {
      const currentTime = new Date();

      if (currentTime > sunrise && currentTime < sunset) {
        toggleTheme = true;
      }
    }

    return toggleTheme;
  };

  private getSunriseAndSunsetTime = async () => {
    const { latitude, longitude } = await this.getLatitudeLongitude();

    if (isNaN(latitude)) {
      Snackbar.show({
        text:
          "To use the automatic theme feature, location services must be turned on. Turning off automatic theme."
      });
      return { sunset: null, sunrise: null };
    }

    const sunset = getSunset(latitude, longitude);
    const sunrise = getSunrise(latitude, longitude);
    return { sunset, sunrise };
  };

  private async getLatitudeLongitude() {
    const currentDate = moment().toISOString();
    const lastQueried = await AsyncStorage.getItem("@LastQueriedLocation");
    let lastQueriedDate = currentDate;

    if (lastQueried !== null) {
      lastQueriedDate = moment(lastQueried)
        .add(7, "days")
        .toISOString();
    }

    let latitude = parseInt(
      (await AsyncStorage.getItem("@Latitude")) || "0",
      10
    );
    let longitude = parseInt(
      (await AsyncStorage.getItem("@Longitude")) || "0",
      10
    );

    if (lastQueriedDate > currentDate) {
      ({ latitude, longitude } = await this.getNewLatitudeLongitude(
        currentDate,
        latitude,
        longitude
      ));
    }
    return { latitude, longitude };
  }

  private async getNewLatitudeLongitude(
    currentDate: string,
    latitude: number,
    longitude: number
  ) {
    const { status } = await Permissions.askAsync(Permissions.LOCATION);
    if (status === "granted") {
      const locationOn = await Location.hasServicesEnabledAsync();
      if (locationOn) {
        const location = await Location.getCurrentPositionAsync({
          accuracy: Location.Accuracy.Lowest
        });

        latitude = location.coords.latitude;
        longitude = location.coords.longitude;
        await AsyncStorage.setItem("@Latitude", JSON.stringify(latitude));
        await AsyncStorage.setItem("@Longitude", JSON.stringify(longitude));
        await AsyncStorage.setItem("@LastQueriedLocation", currentDate);
      }
    }
    return { latitude, longitude };
  }
}
