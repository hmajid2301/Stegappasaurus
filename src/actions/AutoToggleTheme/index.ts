import AutoToggleTheme from "./AutoToggleTheme";

export default AutoToggleTheme;
