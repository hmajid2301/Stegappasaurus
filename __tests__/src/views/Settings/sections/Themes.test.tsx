import { mount, shallow } from "enzyme";
import React from "react";
import configureStore, { MockStoreEnhanced } from "redux-mock-store";

import AutoToggleTheme from "~/actions/AutoToggleTheme";
import ThemesConnected, { Themes } from "~/views/Settings/sections/Themes";

describe("Themes: Match Snapshots", () => {
  test("1", () => {
    const component = shallow(
      <Themes
        isAutomatic={true}
        theme={{ color: "#17212D", background: "#FFF", isDark: false }}
        toggleAutomaticTheme={(isAutomatic: boolean) => null}
        toggleDarkTheme={(isDark: boolean) => null}
      />
    );
    expect(component).toMatchSnapshot();
  });

  test("2", () => {
    const component = shallow(
      <Themes
        isAutomatic={true}
        theme={{ color: "#FFF", background: "#17212D", isDark: false }}
        toggleAutomaticTheme={(isAutomatic: boolean) => null}
        toggleDarkTheme={(isDark: boolean) => null}
      />
    );
    expect(component).toMatchSnapshot();
  });
});

describe("Themes: Props", () => {
  test("OnPress: Automatic", async () => {
    const toggle = jest.fn();
    const component = mount(
      <Themes
        isAutomatic={true}
        theme={{ color: "#FFF", background: "#17212D", isDark: false }}
        toggleAutomaticTheme={(isAutomatic: boolean) => null}
        toggleDarkTheme={() => toggle()}
      />
    );

    const spy = jest
      .spyOn(AutoToggleTheme.prototype, "shouldToggleDarkTheme")
      .mockResolvedValue(true);

    const prop = component
      .find("TouchableOpacity")
      .last()
      .props() as any;
    await prop.onPress();

    expect(spy).toHaveBeenCalled();
    expect(toggle).toHaveBeenCalled();
  });
});

describe("Themes: Redux", () => {
  let store: MockStoreEnhanced<unknown, {}>;
  beforeAll(() => {
    const mockStore = configureStore();
    const initialState = {
      ToggleAutomaticTheme: {
        isAutomatic: false
      }
    };
    store = mockStore(initialState);
  });

  test("mapStateToProps", async () => {
    const component = shallow(
      //@ts-ignore
      <ThemesConnected store={store} />
    ).dive();
    expect((component.props() as any).isAutomatic).toEqual(false);
  });

  test("mapDispatchToProps", () => {
    const component = shallow(
      //@ts-ignore
      <ThemesConnected store={store} />
    ).dive();
    (component.props() as any).toggleAutomaticTheme(true);
    (component.props() as any).toggleDarkTheme(true);
    const actions = store.getActions();
    expect(actions).toMatchSnapshot();
  });
});
