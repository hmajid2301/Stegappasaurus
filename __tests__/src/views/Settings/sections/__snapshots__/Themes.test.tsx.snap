// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Themes: Match Snapshots 1 1`] = `
<View>
  <Themed.ListItem
    containerStyle={
      Object {
        "backgroundColor": "#FFF",
      }
    }
    title={
      <Text
        style={
          Object {
            "color": "#009CFF",
            "fontFamily": "Roboto",
            "fontSize": 12,
          }
        }
      >
        Themes
      </Text>
    }
    titleStyle={
      Object {
        "paddingBottom": 5,
        "paddingTop": 20,
      }
    }
  />
  <Themed.ListItem
    bottomDivider={true}
    checkBox={
      Object {
        "checked": false,
        "checkedColor": "#009CFF",
        "onPress": [Function],
      }
    }
    containerStyle={
      Object {
        "backgroundColor": "#FFF",
      }
    }
    title="Dark Mode"
    titleStyle={
      Array [
        Object {
          "fontFamily": "Roboto-Light",
        },
        Object {
          "color": "#17212D",
        },
      ]
    }
    topDivider={true}
  />
</View>
`;

exports[`Themes: Match Snapshots 2 1`] = `
<View>
  <Themed.ListItem
    containerStyle={
      Object {
        "backgroundColor": "#17212D",
      }
    }
    title={
      <Text
        style={
          Object {
            "color": "#009CFF",
            "fontFamily": "Roboto",
            "fontSize": 12,
          }
        }
      >
        Themes
      </Text>
    }
    titleStyle={
      Object {
        "paddingBottom": 5,
        "paddingTop": 20,
      }
    }
  />
  <Themed.ListItem
    bottomDivider={true}
    checkBox={
      Object {
        "checked": false,
        "checkedColor": "#009CFF",
        "onPress": [Function],
      }
    }
    containerStyle={
      Object {
        "backgroundColor": "#17212D",
      }
    }
    title="Dark Mode"
    titleStyle={
      Array [
        Object {
          "fontFamily": "Roboto-Light",
        },
        Object {
          "color": "#FFF",
        },
      ]
    }
    topDivider={true}
  />
</View>
`;
